fsspec/asyn.py

```diff
@@ -0,0 +1,206 @@
+
+    @functools.wraps(func)
+    async def wrapper(*args, **kwargs):
+        return func(*args, **kwargs)
+
+    return wrapper
+
+
+def get_loop():
+    """Create a running loop in another thread"""
+    loop = asyncio.new_event_loop()
+    t = threading.Thread(target=loop.run_forever)
+    t.daemon = True
+    t.start()
+    return loop
+
+
+# these methods should be implemented as async by any async-able backend
+async_methods = [
+    "_ls",
+    "_expand_path",
+    "_info",
+    "_isfile",
+    "_isdir",
+    "_exists",
+    "_walk",
+    "_glob",
+    "_find",
+    "_du",
+    "_cat",
+    "_get_file",
+    "_put_file",
+    "_rm_file",
+]
+
+
+class AsyncFileSystem:
+    async_impl = True
+
+    async def _rm(self, path, recursive=False):
+        path = await self._expand_path(path, recursive=recursive)
+        await asyncio.gather(*[self._rm_file(p) for p in path])
+
+    async def _mcat(self, paths, recursive=False):
+        """Fetch multiple paths' contents"""
+        paths = await self._expand_path(paths, recursive=recursive)
+        out = await asyncio.gather(
+            *[asyncio.ensure_future(self._cat(path), loop=self.loop) for path in paths]
+        )
+        return {k: v for k, v in zip(paths, out)}
+
+    async def _put(self, lpath, rpath, recursive=False, **kwargs):
+        """Copy file(s) from local.
+
+        Copies a specific file or tree of files (if recursive=True). If rpath
+        ends with a "/", it will be assumed to be a directory, and target files
+        will go within.
+
+        Calls put_file for each source.
+        """
+        from .implementations.local import make_path_posix, LocalFileSystem
+
+        rpath = self._strip_protocol(rpath)
+        if isinstance(lpath, str):
+            lpath = make_path_posix(lpath)
+        fs = LocalFileSystem()
+        lpaths = fs.expand_path(lpath, recursive=recursive)
+        rpaths = other_paths(lpaths, rpath)
+
+        await asyncio.gather(
+            *[
+                self._put_file(lpath, rpath, **kwargs)
+                for lpath, rpath in zip(lpaths, rpaths)
+            ]
+        )
+
+    async def _get(self, rpath, lpath, recursive=False, **kwargs):
+        """Copy file(s) to local.
+
+        Copies a specific file or tree of files (if recursive=True). If lpath
+        ends with a "/", it will be assumed to be a directory, and target files
+        will go within.
+
+        Calls get_file for each source.
+        """
+        from fsspec.implementations.local import make_path_posix
+
+        rpath = self._strip_protocol(rpath)
+        lpath = make_path_posix(lpath)
+        rpaths = await self._expand_path(rpath, recursive=recursive)
+        lpaths = other_paths(rpaths, lpath)
+        await asyncio.gather(
+            *[
+                self._get_file(rpath, lpath, **kwargs)
+                for lpath, rpath in zip(lpaths, rpaths)
+            ]
+        )
+
+
+def make_sync_methods(obj):
```

Comment:
So this _also_ create async methods right ? 

1) Do you think a better like "mirror" methods would be better; 
and
2) do you think some `obj` would implement both the sync and async in which case IIUC this would overwrite the custom impl ?

Reply:
Agreed, the name could be better (obviously the func first did only one of the jobs...).
Yes, you would squash the sync implementation; it could apply only if the method being overwritten is missing of the same as the AsyncFileSystem class.